README.md
# Blackjs
> Blackjs is a JavaScript Library to ease the way of programming

## Importation
In your HTML file add the black.js script ```<script src ="blackjs.js"></script>```

## Develop
```
npm install
```
## Build
Run that code at the terminal
```
grunt
```
> It generates a build and a minifier of the library
or use
```
grunt watch
```
to keep tracking the files while you are changing it



CHANGELOG.md

#4.1.0 (9/1/2020) [important]

* Added godaneCode
* Added ruleOfThree() function
* Added element.get() function to use instread of getElement()
* Added deprecated warning to getElement() function
* The contain() function allows you to decide to differentiate between capital letters
* The go() function now is called navigate() function
* The loop() function sintaxis has changed
* Grunt watch Added
* Grunt problems and vulnerabilities solved

#4.0.0 (17/6/2018) [important]

* Added Grunt support

#3.1.0 (2/2/2018) [important]

* Removed insert() function because it was deprecated
* Added getElement().html.add() function
* Added getElement().html.set() function
* Added styles.write() function

#3.0.0 (2/2/2018) [important]

* Removed getText() function because it was deprecated
* Removed getImageSrc() function because it was deprecated
* Removed setImageSrc() function because it was deprecated
* Removed onPressKey() event because it was deprecated
* Removed onClick() event because it was deprecated
* Removed onChange() event because it was deprecated
* Removed anim() function because it was deprecated

#2.5.2 (23/1/2018)

* Removed JavaScript-MD5
* Removed js-sha3
* Removed js-sha1
* The contains() function doesn't differentiate between capital letters

#2.5.1 (13/1/2018)

* Added JavaScript-MD5

#2.5.0 (7/1/2018)

* Added js-sha3
* Added js-sha1
* Changed the sha1.encrypt(text) function to sha1(text)
* Changed insert() function

#2.4.0 (26/12/2017)

* Added random() function
* Added randomize() function

#2.3.0 (25/12/2017)

* Added a callback to the countdown() function
* Added deprecated warning to anim() function
* Added deprecated warning to onClick() function
* Added deprecated warning to onPressKey() function
* Added deprecated warning to onChange() function
* Changed anim function to getElement().anim()
* Changed onClick() event to getElement().onClick()
* Changed onPressKey() event to getElement().onPressKey()
* Changed onChange() event to getElement().onChange()

#2.2.0 (25/12/2017)

* Added deprecated warning to setImageSrc() function
* Added deprecated warning to getImageSrc() function
* Added deprecated warning to getText() function
* Changed base64('encrypt',msg) function to base64.encrypt(msg)
* Changed base64('decrypt',msg) function to base64.decrypt(msg)
* Changed sha1(msg) function to sha1.encrypt(msg)


#2.1.0 (25/12/2017)

* Added getElement().width() function
* Added getElement().height() function
* Added getElement().element() function
* Added support to change the text of an element via getElement().text(text)
* Added support to change the width of an element via getElement().width(width)
* Added support to change the height of an element via getElement().height(height)
* Added support to change the src of an image via getElement().src(src)


#2.0.1 (24/12/2017)

* Fixed getElement().src() function
* Fixed getElement().text() function


#2.0.0 (24/12/2017)

* Changed the startTime function to start.time()
* Changed the stopTime function to stop.time()
* Changed the use for getElement() function
* Added getElement().text() function
* Added getElement().src() function

#1.6.0 (23/12/2017)

* Added startTimer() function
* Added stopTimer() function
* Added clearData() function
* Added countdown() function

#1.5.0 (23/12/2017)

* Added base64() function
* Added anim() function
* Added anim('#','slide','top') support
* Added anim('#','slide','bottom') support
* Added anim('#','slide','right') support
* Added anim('#','slide','left') support
* Added anim('#','enlarge','right') support
* Added anim('#','enlarge','bottom') support
* Added anim('#','fade','in') support
* Added anim('#','fade','out') support


#1.4.0 (23/12/2017)

* Added wait() function
* Added sha1() function
* Added toString() function
* Added toInt() function
* Added doTTT() function
* Added insert('ul') support
* Added insert('li') support

#1.3.1 (22/12/2017) [important]

* Fixed "other" problem
* Fixed classes problem

#1.3.0 (22/12/2017)

* Added saveData() function
* Added getData() function

#1.2.0 (22/12/2017) 72014a6

* Added insert('h5') support
* Added insert('h6') support
* Added insert('div') support
* Added isEqual() function
* Added contains() function

#1.1.0 (22/12/2017) bd762ab

* Initial Commit


blackjs.js

// Copyright (c) 2017 Joaquin Giordano
// version: 3.1.0


function checkIdClassOther(text){
	var fl = text.charAt(0);
	if (fl == "#"){
		return "id";
	}
	if (fl == "."){
		return "class";
	}
	if (fl != "#" && fl != "."){
		return "other";
	}
}



var styles = {

		add:function(text){
					document.head.innerHTML += '<style> '+ text + '</style>';
    			}

};


function log(text){
	console.log(text);
}


function getElement(idclass){
	var checkIdClass = checkIdClassOther(idclass);

	if(checkIdClass == "id"){
		idclass = idclass.substr(1);
		idclass = document.getElementById(idclass);
	}
	if(checkIdClass == "class"){
		idclass = idclass.substr(1);
		idclass = document.getElementsByClassName(idclass)[0];
	}
	if(checkIdClass == "other"){
		if(idclass == "body"){
			idclass = document.body;
		}

	}

	function text(text){
		if(text != undefined){

			if(idclass.value != undefined){
				idclass.value = text;
			}else{
				idclass.innerHTML = text;

			}

		}else{
			if(idclass.value != undefined){
				return idclass.value;
			}else{
				return idclass.innerHTML;

			}
		}


	}

function src(src){
		if(src != undefined){
			idclass.src = src;
		}else{
			return idclass.src;
		}

	}

function width(px){

		if(px != undefined){
			 idclass.style.width = px ;
		}else{
			 return idclass.offsetWidth + "px";
		}

}

function height(px){
	if(px != undefined){
		 idclass.style.height = px ;
	}else{
		 return idclass.offsetHeight + "px";
	}
}

var html = {
	add: function(text){
		idclass.innerHTML += text;
	},
	set: function(text){
		idclass.innerHTML = text;
	}

}

function anim(type,to,pos,quantity,velocity = 1,callback) {
    idclass.style.display = "block";

    if(type == "slide"){
        if(to == "left"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos == quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos++;
                    idclass.style.right = pos + 'px';
                }
            }
        }
        if(to == "right"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos == quantity) {
                    clearInterval(id);
                    if(callback != undefined){
                        callback();
                    }
                } else {
                    pos++;
                    idclass.style.left = pos + 'px';
                }
            }
        }
        if(to == "top"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos == quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos++;
                    idclass.style.bottom = pos + 'px';
                }
            }
        }

        if(to == "bottom"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos == quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos++;
                    idclass.style.top = pos + 'px';
                }
            }
        }
      }

      if(type == "enlarge"){
        if(to == "bottom"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos == quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos++;
                    idclass.style.height = pos + 'px';
                }
            }
        }
        if(to == "right"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos == quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos++;

                    idclass.style.width = pos + 'px';
                }
            }
        }
      }
      if(type == "fade"){
        if(to == "in"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos >= quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos += 0.002;

                    idclass.style.opacity = pos ;
                }
            }
        }
        }
        if(to == "out"){
             var id = setInterval(frame, velocity);
             function frame() {
                if (pos <= quantity) {
                    clearInterval(id);
                     if(callback != undefined){
                        callback();
                    }
                } else {
                    pos -= 0.002;

                    idclass.style.opacity = pos ;
                }
            }
        }
      }


			function onPressKey(letter,callback) {

				var letter = letter;

					idclass.onkeypress = function(e,letter2 = letter){

					if(e.which == letter2){
						callback();
					}

				}

			}

			function onClick(callback) {

					idclass.onclick = function(){
						callback();

				}
			}

			function onChange(callback) {

					idclass.onchange = function(){
						callback();
					}
			}

function element(){
	return idclass;
}

		return{
			text:text,
			height:height,
			width:width,
			src:src,
			anim:anim,
			html:html,
			onClick:onClick,
			onChange:onChange,
			onPressKey:onPressKey,
			element:element

		};


}



	function random(num1,num2){

		if(num1 == undefined || num2 == undefined){

			return Math.random();

		}else {
			return Math.floor(Math.random() * num2) + num1;
		}

	}

	function randomize(data) {
		var num = random(0,data.length);
		return data[num];
	}

function contains(where,data){
	var checkWhere = checkIdClassOther(where);
	if(checkWhere == "id"){
		where = where.substr(1);
		try{
			var number = document.getElementById(where).value.toLowerCase().indexOf(data.toLowerCase());
			if(number != -1){
				return true;
			}else{
				return false;
			}
		}catch(err){
			var number = document.getElementById(where).innerHTML.toLowerCase().indexOf(data.toLowerCase());
			if(number != -1){
				return true;
			}else{
				return false;
			}
		}
	}
	if(checkWhere == "class"){
		where = where.substr(1);
			try{
				var data2 = document.getElementsByClassName(where)[0].value.toLowerCase().indexOf(data.toLowerCase());
				if(data2 != undefined){
				if(data2 != -1){
					return true;
				}else{
					return false;
				}
			}
			}catch(err){
				var data = document.getElementsByClassName(where)[0].innerHTML.toLowerCase().indexOf(data.toLowerCase());
				if(data != -1){
					return true;
				}else{
					return false;
				}
			}
		}
	if(checkWhere == "other"){
		var number = where.indexOf(data.toLowerCase());
		if(number != -1){
			return true;
		}else{
			return false;
		}
	}

}

function isEqual(data){
	var result = true;
	var fs = true;
	var primary_data;
	data.forEach(function(e){
		if(fs == true){
			primary_data = e;
			fs = false;
		}else{
			if(primary_data != e){
				result = false;

			}
		}
	});
	return result;
}

function sum(numbers){
var fs = true;
	var actualNumber = 0;
	numbers.forEach(function(e) {
		if(fs == true){
			actualNumber = e;
			fs = false;
		}else{
			actualNumber += e;
		}
	})
	return actualNumber;
}

function subs(numbers){
	var fs = true;
	var actualNumber = 0;
	numbers.forEach(function(e) {
		if(fs == true){
			actualNumber = e;
			fs = false;
		}else{
			actualNumber -= e;
		}
	})
	return actualNumber;
}

function devide(numbers){
var fs = true;
	var actualNumber = 0;
	numbers.forEach(function(e) {
		if(fs == true){
			actualNumber = e;
			fs = false;
		}else{
			actualNumber /= e;
		}
	})
	return actualNumber;
}

function multiply(numbers){
var fs = true;
	var actualNumber = 0;
	numbers.forEach(function(e) {
		if(fs == true){
			actualNumber = e;
			fs = false;
		}else{
			actualNumber *= e;
		}
	})
	return actualNumber;
}

// countdown example: countdown('13 aug 2021')
function countdown(date,days_count = true,hours_count = true,minutes_count = true,seconds_count = true,callback){
	var countDownDate = new Date(date).getTime();
  var now = new Date().getTime();
  var distance = countDownDate - now;
  var days = Math.floor(distance / (1000 * 60 * 60 * 24));
  var hours = Math.floor((distance % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
  var minutes = Math.floor((distance % (1000 * 60 * 60)) / (1000 * 60));
  var seconds = Math.floor((distance % (1000 * 60)) / 1000);
  var result;
  var ft = true;
  if (days_count == true){
   if(ft){
     result = days + "d ";
     ft = false;
    }else{
      result += days + "d ";
    }
   }

  if (hours_count == true){
      if(ft){
     result = hours + "h ";
     ft = false;
    }else{
      result += hours + "h ";
    }
  }
  if (minutes_count == true){
      if(ft){
     result = minutes + "m ";
     ft = false;
    }else{
      result += minutes + "m ";
    }
  }
  if (seconds_count == true){
      if(ft){
     result = seconds + "s ";
     ft = false;
    }else{
      result += seconds + "s ";
    }
  }
  if (distance < 0) {
		if(callback == undefined){
			callback();
		}
    return "Expired";
  }else{
    return result;
  }


}

function go(url){
	window.location = url;
}

function clearData(){
	localStorage.clear();
}

function saveData(where,data){
	var checkData = checkIdClassOther(data);
	if(checkData == "id"){
		data = data.substr(1);
		try{
			var data = document.getElementById(data).value;
			localStorage.setItem(where,data);
		}catch(err){
			var data = document.getElementById(data).innerHTML;
			localStorage.setItem(where,data);
		}
	}
	if(checkData == "class"){
		data = data.substr(1);
			var data2 = document.getElementsByClassName(data)[0].value;
			if(data2 != undefined){
				localStorage.setItem(where,data);
			}else{
			var data = document.getElementsByClassName(data)[0].innerHTML;

			localStorage.setItem(where,data);
			}


	}
	if(checkData == "other"){

		localStorage.setItem(where,data);
	}

}

function getData(where){
	return localStorage.getItem(where);
}

function getChar(where,num){
	var checkWhere = checkIdClassOther(where);

	if(checkWhere == "other"){
		where = where;
	}
	if(checkWhere == "id"){
		where = where.substr(1);
		var prewhere = document.getElementById(where).value;
		if(prewhere == undefined){
			where = document.getElementById(where).innerHTML;
		}else{
			where = prewhere;
		}
	}
	if(checkWhere == "class"){
		where = where.substr(1);
		var prewhere = document.getElementsByClassName(where)[0].value;
		if(prewhere == undefined){
			where = document.getElementsByClassName(where)[0].innerHTML;
		}else{
			where = prewhere;
		}
	}
	return where.charAt(num);
}

function doTTT(callback,callback2) {
	callback();
	callback2();
}

function time(data){

	var date = new Date();
	var final = "";
	data.forEach(function(e) {

	if(e == undefined){
		final == date;
	}
	if(e == "hour"){
		if(date.getHours() <= 9){
			final += "0" + date.getHours() + " ";
		}else{
			final += date.getHours() + " ";
		}
	}
	if(e == "minute"){
		if(date.getMinutes() <= 9){
			final += "0" + date.getMinutes() + " ";
		}else{
			final += date.getMinutes() + " ";
		}
	}
	if(e == "second"){
		if(date.getSeconds() <= 9){
			final += "0" + date.getSeconds() + " ";
		}else{
			final += date.getSeconds() + " ";
		}
	}
	if(e == "milisecond"){
		if(date.getMilliseconds() <= 9){
			final += "0" + date.getMilliseconds() + " ";
		}else{
			final += date.getMilliseconds() + " ";
		}
	}
	if(e == "day"){
		final += date.getDate() + " ";
	}
	if(e == "dayoftheweek"){
		final += date.getDay() + 1 + " ";
	}
	if(e == "year"){
		final += date.getFullYear() + " ";
	}
	if(e == "month"){
		final += date.getMonth() + 1  + " ";
	}

	})

	return final;
}

function toString(num){
	return num + "";
}

function toInt(string){
	return parseInt(string);
}

function wait(seconds,callback){
	var ticks = seconds * 1000;
	setTimeout(function(){callback();}, ticks);
}

function loop(times,callback){
	times -= 1;
	for(var i = 0; i <= times; i++){
		callback();
	}
}

function onDocumentReady(callback){
	document.onreadystatechange = () => {
  if (document.readyState === 'complete') {
    	callback();
  }
};
}
